documentation/examples/custom-sd/adapter-usage/main.go

```diff
@@ -0,0 +1,251 @@
+		} else {
+			addr = net.JoinHostPort(node.Address, fmt.Sprintf("%d", node.ServicePort))
+		}
+
+		target := model.LabelSet{model.AddressLabel: model.LabelValue(addr)}
+		labels := model.LabelSet{
+			model.AddressLabel:                   model.LabelValue(addr),
+			model.LabelName(addressLabel):        model.LabelValue(node.Address),
+			model.LabelName(nodeLabel):           model.LabelValue(node.Node),
+			model.LabelName(tagsLabel):           model.LabelValue(tags),
+			model.LabelName(serviceAddressLabel): model.LabelValue(node.ServiceAddress),
+			model.LabelName(servicePortLabel):    model.LabelValue(strconv.Itoa(node.ServicePort)),
+			model.LabelName(serviceIDLabel):      model.LabelValue(node.ServiceID),
+		}
+		tgroup.Labels = labels
+
+		// Add all key/value pairs from the node's metadata as their own labels.
+		for k, v := range node.NodeMeta {
+			name := strutil.SanitizeLabelName(k)
+			tgroup.Labels[model.LabelName(model.MetaLabelPrefix+name)] = model.LabelValue(v)
+		}
+
+		tgroup.Targets = append(tgroup.Targets, target)
+	}
+	return &tgroup, nil
+}
+
+// Note: you must implement this function for your discovery implementation as part of the
+// Discoverer interface. Here you should query your SD for it's list of known targets, determine
+// which of those targets you care about (for example, which of Consuls known services do you want
+// to scrape for metrics), and then send those targets as a target.TargetGroup to the ch channel.
+func (d *discovery) Run(ctx context.Context, ch chan<- []*targetgroup.Group) {
+OUTER:
+	for c := time.Tick(time.Duration(d.refreshInterval) * time.Second); ; {
+		var srvs map[string][]string
+		resp, err := http.Get(fmt.Sprintf("http://%s/v1/catalog/services", d.address))
+
+		if err != nil {
+			level.Error(d.logger).Log("msg", "Error getting services list", "err", err)
+			time.Sleep(time.Duration(d.refreshInterval) * time.Second)
+			continue
+		}
+
+		dec := json.NewDecoder(resp.Body)
+		err = dec.Decode(&srvs)
+		resp.Body.Close()
+		if err != nil {
+			level.Error(d.logger).Log("msg", "Error reading services list", "err", err)
+			time.Sleep(time.Duration(d.refreshInterval) * time.Second)
+			continue
+		}
+
+		var tgs []*targetgroup.Group
+		// Note that we treat errors when querying specific consul services as fatal for for this
+		// iteration of the time.Tick loop. It's better to have some stale targets than an incomplete
+		// list of targets simply because there may have been a timeout. If the service is actually
+		// gone as far as consul is concerned, that will be picked up during the next iteration of
+		// the outer loop.
+		for name := range srvs {
+			if name == "consul" {
+				continue
+			}
+			resp, err := http.Get(fmt.Sprintf("http://%s/v1/catalog/service/%s", d.address, name))
+			if err != nil {
+				level.Error(d.logger).Log("msg", "Error getting services nodes", "service", name, "err", err)
+				goto OUTER
```

Comment:
I think you meant continue rather than goto

Reply:
no I meant `goto`, but the way I'd set up the ticker this won't work

the label for the goto should be at the select statement below

Reply:
I'd really like to avoid having a goto.

If the for is just being used for the initialiser, Id say it'd be cleaner to move the initialisation before the for.

Reply:
I think the thing to do is move the err outside of the inner for, break the inner for if there's an error, and only send to the channel if there is no error.